mlflow/projects/__init__.py

```diff
@@ -515,22 +512,91 @@ def _get_conda_command(conda_env_name):
+
+
+def _get_docker_command(work_dir, dockerimage):
+    docker_path = _get_docker_bin_executable("docker")
+    cmd = [docker_path, "run", "--rm"]
+    cmd += [dockerimage]
+    return cmd
+
+
+def _get_docker_bin_executable(executable_name):
+    """
+    Return path to the specified executable, assumed to be discoverable within the 'bin'
+    subdirectory of a docker installation.
+
+    The docker home directory (expected to contain a 'bin' subdirectory) is configurable via the
+    ``mlflow.projects.MLFLOW_DOCKER_HOME`` environment variable. If
+    ``mlflow.projects.MLFLOW_DOCKER_HOME`` is unspecified, this method simply returns the passed-in
+    executable name.
+    """
+    docker_home = os.environ.get(MLFLOW_DOCKER_HOME)
+    if docker_home:
+        return os.path.join(docker_home, "bin/%s" % executable_name)
+    return executable_name
+
+
+def _validate_docker_installation():
+    """
+    Verify if Docker is installed on host machine.
+    """
     try:
-        process.exec_cmd(["docker", "--help"], throw_on_error=False)
+        docker_path = _get_docker_bin_executable("docker")
+        process.exec_cmd([docker_path, "--help"], throw_on_error=False)
     except EnvironmentError:
         raise ExecutionException("Could not find Docker executable. "
                                  "Ensure Docker is installed as per the instructions "
                                  "at https://docs.docker.com/install/overview/.")
+
+
+def _validate_docker_env(docker_env):
     if not docker_env.get('dockerimage'):
         raise ExecutionException("No docker dockerimage specified.")
 
 
-def _validate_execution_environments(project):
-    if project.conda_env_path and project.docker_env:
-        raise ExecutionException("Both conda and docker specification found. "
-                                 "Please specify just one execution environment in MLProject.")
-
+def _build_docker_image(work_dir, project, active_run, project_mount_path="/mlflow/projects/code"):
+    os.chdir(work_dir)
+    docker_path = _get_docker_bin_executable("docker")
+    built_image = "mlflow-{name}".format(
+        name=(project.name if project.name else "built-image"))
+
+    dockerfile_template = (
+        "FROM {imagename}\n"
+        "LABEL Name={built_image} Version=0.0.1 AUTHOR=MarcusRehm\n"
+        "COPY {work_dir} {project_mount_path}\n"
+        "# RUN pip install mlflow=={version}\n"
+        # "ENTRYPOINT [\"mlflow\", \"run\", \"{project_mount_path}\"] \n"
+    ).format(imagename=project.docker_env.get('dockerimage'), built_image=built_image,
+             work_dir=".", project_mount_path=project_mount_path,
+             version=mlflow.version.VERSION)
+    dockerfile_template_env_vars = []
+    for key, value in _get_run_env_vars(run_id=active_run.info.run_uuid,
+                                        experiment_id=active_run.info.experiment_id).items():
+        dockerfile_template_env_vars += ["ENV {key}={value}\n".format(key=key, value=value)]
```

Comment:
Good, I agree with you @smurching , right now we can validate to run only if the tracking is a remote URI. For docker I think this would be the most common approach.

Reply:
I think the first option would be better for now, don't you think? I'm validating if the tracking URI is local and if so it throws an exception say that it needs to be a remote URI.